### PR TITLE
Create eslint rule suggesting useLayoutEffect from @react-aria

### PR DESCRIPTION
**Pull request by frdclav**
This PR adds an esLint rule for using the @react-aria/utils implementation of useLayoutEffect, and updates any existing instances in the codebase not using it.

Closes 
#1494 

## ✅ Pull Request Checklist:

- [x] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [x] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [x] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

1. Add this import anywhere in the project: ```import { useLayoutEffect } from react```
2. Run ```yarn lint```
3. See that an error is thrown ```error  Please use useLayoutEffect from @react-aria/utils instead  rulesdir/useLayoutEffectRule```

Confirmed all tests are passing after updating existing useLayoutEffect imports:
Screen Shot 2021-12-10 at 10 29 17 AM

### CODE REVIEW

**Comment by frdclav** [comment]
I am new to creating/update esLint rules but I looked into using no-restricted-imports:
```
 'no-restricted-imports': ["error", {
    "paths": [{
    "name": "react",
    "importNames": ["useLayoutEffect"],
    "message": "Please use useLayoutEffect from @react-aria instead."
      }]
    }],
```
With this I was having trouble coming up with the correct option or pattern for matching imports from anything except @react-aria/utils.

I tried going with a custom plugin and this seems to work:
```
module.exports = function (context) {
  return {
    ImportDeclaration(node) {
      const source = node.source.value;
      if (source === '@react-aria/utils' || source === './') {
        return;
      }
      const importSpecifiers = node.specifiers.filter(specifier => specifier.type === 'ImportSpecifier');
      const getName = specifier => specifier.local.name;
      importSpecifiers.map(
                (item) => {
                  let itemName = getName(item);
                  if (itemName === 'useLayoutEffect') {
                    context.report(node, 'Please use useLayoutEffect from @react-aria/utils instead.');
                  }
                });
      return;
    }
  };
};
```

qq: do we need to restrict this rule within @react-aria only, or should it cover everything?

**Comment by snowystinger** [comment]
it should cover everything

**Comment by frdclav** [comment]
Should I also update current instances where useLayoutEffect is not using the implementation from @frdclav/@react-aria/utils?
```
[2] node scripts/lint-packages.js exited with code 0
[0] yarn check-types exited with code 0
[1]
...
[1] /Users/rodsoncl/react-spectrum/packages/@react-aria/dnd/src/useDrop.ts
[1]   13:1  error  Please use useLayoutEffect from @react-aria/utils instead  rulesdir/useLayoutEffectRule
[1]
[1] /Users/rodsoncl/react-spectrum/packages/@react-aria/dnd/src/useDroppableCollection.ts
[1]   17:1  error  Please use useLayoutEffect from @react-aria/utils instead  rulesdir/useLayoutEffectRule
...
[1] /Users/rodsoncl/react-spectrum/packages/@react-aria/ssr/src/SSRProvider.tsx
[1]   13:1   error    Please use useLayoutEffect from @react-aria/utils instead                                             rulesdir/useLayoutEffectRule
...
[1] /Users/rodsoncl/react-spectrum/packages/@react-aria/tabs/src/useTabPanel.ts
[1]   16:1  error  Please use useLayoutEffect from @react-aria/utils instead  rulesdir/useLayoutEffectRule
[1]
[1] /Users/rodsoncl/react-spectrum/packages/@react-aria/utils/src/useDescription.ts
[1]   14:1  error  Please use useLayoutEffect from @react-aria/utils instead  rulesdir/useLayoutEffectRule
[1]
[1] /Users/rodsoncl/react-spectrum/packages/@react-aria/utils/src/useId.ts
[1]   14:1  error    Please use useLayoutEffect from @react-aria/utils instead                                                                    ...
[1] /Users/rodsoncl/react-spectrum/packages/@react-aria/utils/test/useObjectRef.test.js
[1]   13:1  error  Please use useLayoutEffect from @react-aria/utils instead  rulesdir/useLayoutEffectRule
[1]
...
[1] /Users/rodsoncl/react-spectrum/packages/@react-spectrum/autocomplete/src/SearchAutocomplete.tsx
  27:1  error  Please use useLayoutEffect from @react-aria/utils instead  rulesdir/useLayoutEffectRule
[1]
...
[1] /Users/rodsoncl/react-spectrum/packages/@react-spectrum/searchwithin/src/SearchWithin.tsx
[1]   16:1  error  Please use useLayoutEffect from @react-aria/utils instead  rulesdir/useLayoutEffectRule
[1]
...
[1]
[1] ✖ 52 problems (9 errors, 43 warnings)
[1]
[1] eslint packages --ext .js,.ts,.tsx exited with code 1
error Command failed with exit code 1.
```

**Comment by snowystinger** [comment]
yep, that'd be great!

File: bin/useLayoutEffectRule.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+      const importSpecifiers = node.specifiers.filter(specifier => specifier.type === 'ImportSpecifier');
+      const getName = specifier => specifier.local.name;
+      importSpecifiers.map(
+                (item) => {
+                  let itemName = getName(item);
+                  if (itemName === 'useLayoutEffect') {
+                    context.report(node, 'Please use useLayoutEffect from @react-aria/utils instead.');
+                  }
+                });
+      return;
```
Comment (LFDanLu) [inline bin/useLayoutEffectRule.js:29]: Small formatting nit:
```suggestion
      importSpecifiers.map(
        (item) => {
          let itemName = getName(item);
          if (itemName === 'useLayoutEffect') {
            context.report(node, 'Please use useLayoutEffect from @react-aria/utils instead.');
          }
        }
      );
```

**Review by LFDanLu — APPROVED** [review]
LGTM, confirmed that this rule caught `import {useLayoutEffect} from 'react'` in each of our packages when testing locally.

**Review by snowystinger — APPROVED** [review]
looks good!

**Pull request merged**